pkg/steps/template_test.go

```diff
@@ -0,0 +1,173 @@
+package steps
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/openshift/ci-operator/pkg/api"
+	"k8s.io/apimachinery/pkg/util/diff"
+)
+
+func TestDeferredParametersAllLinks(t *testing.T) {
+	var testCases = []struct {
+		purpose       string
+		dp            *DeferredParameters
+		expectedItems int
+	}{{
+		purpose: "sanity",
+		dp: &DeferredParameters{
+			links: map[string][]api.StepLink{
+				"K1": {someStepLink("ONE"), someStepLink("TWO")},
+				"K2": {someStepLink("THREE")},
+			},
+		},
+		// only compare count of returned items because AllLinks can
+		// return links in any order
+		expectedItems: 3,
+	}}
+
+	for _, tc := range testCases {
+		links := tc.dp.AllLinks()
+
+		if len(links) != tc.expectedItems {
+			t.Errorf("%s: %v.AllLinks() returned %v, expected %d items", tc.purpose, tc.dp, links, tc.expectedItems)
+		}
+	}
+}
+
+func TestDeferredParametersMap(t *testing.T) {
+	var testCases = []struct {
+		purpose  string
+		dp       *DeferredParameters
+		expected map[string]string
+	}{{
+		purpose: "sanity",
+		dp: &DeferredParameters{
+			values: map[string]string{"K1": "V1", "K2": "V2"},
+			fns: map[string]func() (string, error){
+				"K1": func() (string, error) { return "F1", nil },
+				"K3": func() (string, error) { return "F3", nil },
+			},
+		},
+		expected: map[string]string{"K1": "V1", "K3": "F3"},
+	}}
+
+	for _, tc := range testCases {
+		createdMap, _ := tc.dp.Map()
+
+		if !reflect.DeepEqual(tc.expected, createdMap) {
+			t.Errorf("%s: %v.Map() returned different map:\n%s", tc.purpose, tc.dp, diff.ObjectReflectDiff(tc.expected, createdMap))
+		}
+	}
+}
+
+func TestDeferredParametersAddHasLinks(t *testing.T) {
+	var testCases = []struct {
+		purpose string
+		dp      *DeferredParameters
+		name    string
+		link    api.StepLink
+		fn      func() (string, error)
+		has     bool
+		links   []api.StepLink
+	}{{
+		purpose: "New key",
+		dp:      NewDeferredParameters(),
+		name:    "key",
+		link:    someStepLink("name"),
+		fn:      nil,
+		has:     true,
+		links:   []api.StepLink{someStepLink("name")},
+	}, {
+		purpose: "Nonexistent key",
+		dp:      NewDeferredParameters(),
+		name:    "DO-NOT-ADD",
+		link:    nil,
+		fn:      nil,
+		has:     false,
+		links:   nil,
+	}, {
+		purpose: "Existing key",
+		dp: &DeferredParameters{
+			fns: api.ParameterMap{"key": func() (string, error) {
+				return "old", nil
+			}},
+			values: map[string]string{},
+			links:  map[string][]api.StepLink{"key": {someStepLink("old-link")}},
+		},
+		name:  "key",
+		link:  someStepLink("new-link"),
+		fn:    func() (string, error) { return "new", nil },
+		has:   true,
+		links: []api.StepLink{someStepLink("new-link")},
+	}}
+	for _, tc := range testCases {
+		if tc.name != "DO-NOT-ADD" {
```

Comment:
This smells of a bad test array setup?

Reply:
Indeed :)

Reply:
Replaced with explicit `callAdd` / `callSet` booleans.